colony_picker/inverse_kinematics.py

```diff
@@ -0,0 +1,634 @@
+    """Conducts inverse kinematics to find joint angles that result in the
+    desired_end_effector_pose.
+
+    Args:
+        thetas_init (np.ndarray): a 1 x n array containing joint angles to seed
+            the optimization algorithm with in the format
+            [joint angle 1, joint angle 2, ... , joint angle n], where n is the
+            number of joints in the robot arm. The closer these angles are to 
+            the true joint angles, the more likely it is that the optimization
+            will find the true joint angles that lead to the desired end
+            effector pose. 
+        desired_end_effector_pose (np.ndarray): a 1 x 7 array describing the
+            desired position and orientation of the end effector in the format
+            [x position, y position, z position, x, y, z, w] in which the first
+            three values describe the desired position of the end effector and
+            the last four values describe its orientation as a quaternion. 
+        dh_params (np.ndarray): a 4 x n array containing the dh parameters to 
+            describe every joint on the robot arm in the format
+            [[theta offset 1, theta offset 2, ... , theta offset n],
+            [alpha 1, alpha 2, ... , alpha n],
+            [a 1, a 2, ... , a n], 
+            [d 1, d 2, ... , d n]]
+            where n is the number of joints on the robot arm.
+        smart_seed (bool, optional): whether the algorithm should randomly seed
+            the optimization until an error lower than the tolerance is achieved
+            if the error is too high. Defaults to False.
+        tolerance (float, optional): The error tolerated between the desired
+            end effector pose and the end effector pose calculated from the
+            joint values the optimization solved for. Defaults to 1e-06.
+        seeding_max_iters (int, optional): the number of times to optimize again
+            with a new random set of thetas_init before giving up on achieving
+            the error tolerance. Defaults to 100.
+
+    Returns:
+        Tuple(np.ndarray, bool): the set of joint angles the inverse kinematics
+            algorithm solved for and whether the joint angles achieve the 
+            desired end effector pose within the error tolerance in the format
+            (joint angles, whether algorithm achieved error within tolerance)
+    """
+    solved = False
+    result = minimize(objective_function, thetas_init,
+                      (desired_end_effector_pose, dh_params,), method='BFGS')
+    error = result.fun
+    if error <= tolerance:
+        solved = True
+    joint_angles = result.x
+
+    # trying new random seeds until error is within tolerance
+    if not solved:
+        if smart_seed:
+            num_joints = len(dh_params[0])
+            for i in range(seeding_max_iters):
+                new_seed = np.array([random.uniform(-math.pi, math.pi)
```

Comment:
You could use:

`2*np.pi*(np.random.rand(num_joints)-0.5)`

To replace the whole list comprehension